internal/controller/storageclient_controller.go

```diff
@@ -55,54 +64,143 @@ const (
+	if err := mgr.GetCache().IndexField(ctx, &corev1.PersistentVolume{}, pvClusterIDIndexName, func(o client.Object) []string {
+		pv := o.(*corev1.PersistentVolume)
+		if pv != nil &&
+			pv.Spec.CSI != nil &&
+			slices.Contains(csiDrivers, pv.Spec.CSI.Driver) &&
+			pv.Spec.CSI.VolumeAttributes["clusterID"] != "" {
+			return []string{pv.Spec.CSI.VolumeAttributes["clusterID"]}
+		}
+		return nil
+	}); err != nil {
+		return fmt.Errorf("unable to set up FieldIndexer for PV cluster id: %v", err)
+	}
+	if err := mgr.GetCache().IndexField(ctx, &snapapi.VolumeSnapshotContent{}, vscClusterIDIndexName, func(o client.Object) []string {
+		vsc := o.(*snapapi.VolumeSnapshotContent)
+		if vsc != nil &&
+			slices.Contains(csiDrivers, vsc.Spec.Driver) &&
+			vsc.Status != nil &&
+			vsc.Status.SnapshotHandle != nil {
+			parts := strings.Split(*vsc.Status.SnapshotHandle, "-")
+			if len(parts) == 9 {
+				// second entry in the volumeID is clusterID which is unique across the cluster
+				return []string{parts[2]}
+			}
+		}
+		return nil
+	}); err != nil {
+		return fmt.Errorf("unable to set up FieldIndexer for VSC csi driver name: %v", err)
+	}
+
+	if r.AvailableCrds[VolumeGroupSnapshotClassCrdName] {
+		if err := mgr.GetCache().IndexField(ctx, &groupsnapapi.VolumeGroupSnapshotContent{}, vgscClusterIDIndexName, func(o client.Object) []string {
+			vgsc := o.(*groupsnapapi.VolumeGroupSnapshotContent)
+			if vgsc != nil &&
+				slices.Contains(csiDrivers, vgsc.Spec.Driver) &&
+				vgsc.Status != nil &&
+				vgsc.Status.VolumeGroupSnapshotHandle != nil {
+				parts := strings.Split(*vgsc.Status.VolumeGroupSnapshotHandle, "-")
+				if len(parts) == 9 {
+					// second entry in the volumeID is clusterID which is unique across the cluster
+					return []string{parts[2]}
+				}
+			}
+			return nil
+		}); err != nil {
+			return fmt.Errorf("unable to set up FieldIndexer for VGSC csi driver name: %v", err)
+		}
+	}
+
+	if err := mgr.GetCache().IndexField(ctx, &csiopv1a1.ClientProfile{}, ownerUIDIndexName, func(obj client.Object) []string {
 		refs := obj.GetOwnerReferences()
-		var owners []string
+		owners := []string{}
 		for i := range refs {
 			owners = append(owners, string(refs[i].UID))
 		}
 		return owners
 	}); err != nil {
-		return fmt.Errorf("unable to set up FieldIndexer for StorageClaim's owner uid: %v", err)
+		return fmt.Errorf("unable to set up FieldIndexer for client profile owner: %v", err)
 	}
 
+	enqueueStorageClientRequest := handler.EnqueueRequestsFromMapFunc(
+		func(ctx context.Context, _ client.Object) []reconcile.Request {
+			storageClients := &v1alpha1.StorageClientList{}
+			if err := r.Client.List(ctx, storageClients); err != nil {
+				return []reconcile.Request{}
+			}
+
+			request := []reconcile.Request{}
+			for idx := range storageClients.Items {
+				request = append(request, reconcile.Request{NamespacedName: client.ObjectKeyFromObject(&storageClients.Items[idx])})
+			}
+			return request
```

Comment:
Why are we queueing an event for all storage clients?

Reply:
this is enqueued only when the vgsc crd becomes available and all clients need to fetch for vgsc.